cluster-api/pkg/controller/machine/reconcile_test.go

```diff
@@ -0,0 +1,208 @@
+			instanceExists:         true,
+			isDeleting:             true,
+			withFinalizer:          true,
+			numExpectedDeleteCalls: 1,
+		},
+		{
+			name:                   "Delete machine, instance exists without finalizer",
+			objExists:              true,
+			instanceExists:         true,
+			isDeleting:             true,
+			withFinalizer:          false,
+		},
+		{
+			name:                   "Delete machine, instance does not exist, with finalizer",
+			objExists:              true,
+			instanceExists:         false,
+			isDeleting:             true,
+			withFinalizer:          true,
+			numExpectedDeleteCalls: 1,
+		},
+		{
+			name:                   "Delete machine, instance does not exist, without finalizer",
+			objExists:              true,
+			instanceExists:         false,
+			isDeleting:             true,
+			withFinalizer:          false,
+		},
+		{
+			name:                   "Delete machine, skip master",
+			objExists:              true,
+			instanceExists:         true,
+			isDeleting:             true,
+			withFinalizer:          true,
+			isMaster:               true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			machineToTest := getMachine("bar", test.isDeleting, test.withFinalizer, test.isMaster)
+			knownObjects := []runtime.Object{}
+			if test.objExists {
+				knownObjects = append(knownObjects, machineToTest)
+			}
+
+			fakeClient := fake.NewSimpleClientset(knownObjects...)
+			fakeMachineClient := fakeClient.Cluster().Machines(metav1.NamespaceDefault)
+
+			// When creating a new object, it should invoke the reconcile method.
+			cluster := testutil.GetVanillaCluster()
+			cluster.Name = "cluster-1"
+			if _, err := fakeClient.ClusterV1alpha1().Clusters(metav1.NamespaceDefault).Create(&cluster); err != nil {
+				t.Fatal(err)
+			}
+
+			actuator := &CountActuator{ExistsValue: test.instanceExists}
+
+			target := &MachineControllerImpl{}
+			target.actuator = actuator
+			target.clientSet = fakeClient
+			target.machineClient = fakeMachineClient
+
+			var err error
+			err = target.Reconcile(machineToTest)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if actuator.CreateCallCount != test.numExpectedCreateCalls {
+				t.Errorf("Got %v create calls, expected %v", actuator.CreateCallCount, test.numExpectedCreateCalls)
+			}
+			if actuator.DeleteCallCount != test.numExpectedDeleteCalls {
+				t.Errorf("Got %v delete calls, expected %v", actuator.DeleteCallCount, test.numExpectedDeleteCalls)
+			}
+			if actuator.UpdateCallCount != test.numExpectedUpdateCalls {
+				t.Errorf("Got %v update calls, expected %v", actuator.UpdateCallCount, test.numExpectedUpdateCalls)
+			}
+			if actuator.ExistsCallCount != test.numExpectedExistsCalls {
+				t.Errorf("Got %v exists calls, expected %v", actuator.ExistsCallCount, test.numExpectedExistsCalls)
+			}
+
+		})
+	}
+}
+
+func getMachine(name string, isDeleting, hasFinalizer, isMaster bool) *v1alpha1.Machine {
```

Comment:
(nit)
Could be something like m since the scope of the variable is so small.

Reply:
done